cpu/mips_pic32_common/periph/gpio.c

```diff
@@ -65,6 +74,117 @@ static inline int _pin_num(gpio_t pin)
     return (pin & 0x0f);
 }
 
+#ifdef MODULE_PERIPH_GPIO_IRQ
+/**
+ * @brief   The PIC32 family has 7 I/O ports and 16 I/O per port
+ */
+#define PORT_NUMOF          (7U)
+#define GPIO_NUMOF          (16U)
+
+static gpio_flank_t isr_flank[PORT_NUMOF][GPIO_NUMOF];
+static gpio_isr_ctx_t isr_ctx[PORT_NUMOF][GPIO_NUMOF];
+
+#if defined(CPU_FAM_PIC32MX)
+static void isr_handler(uint32_t port_addr)
+{
+    uint32_t cnstat = CNSTATx(_port(port_addr));
+    uint32_t port = PORTx(_port(port_addr));
+    unsigned int pin = 0;
+
+    for (pin = 0; pin < GPIO_NUMOF; ++pin) {
+        if (!(cnstat & (1U << pin)))
+            continue;
+
+        if (isr_flank[_port_num(port_addr)][pin] == GPIO_BOTH
+        || (isr_flank[_port_num(port_addr)][pin] == GPIO_RISING && (port & (1U << pin)))
+        || (isr_flank[_port_num(port_addr)][pin] == GPIO_FALLING && !(port & (1U << pin))))
+            isr_ctx[_port_num(port_addr)][pin].cb(isr_ctx[_port_num(port_addr)][pin].arg);
+    }
+}
+
+static void cn_isr(void)
+{
+#ifdef _PORTA_BASE_ADDRESS
+    isr_handler(_PORTA_BASE_ADDRESS);
+#endif
+#ifdef _PORTB_BASE_ADDRESS
+    isr_handler(_PORTB_BASE_ADDRESS);
+#endif
+#ifdef _PORTC_BASE_ADDRESS
+    isr_handler(_PORTC_BASE_ADDRESS);
+#endif
+#ifdef _PORTD_BASE_ADDRESS
+    isr_handler(_PORTD_BASE_ADDRESS);
+#endif
+#ifdef _PORTE_BASE_ADDRESS
+    isr_handler(_PORTE_BASE_ADDRESS);
+#endif
+#ifdef _PORTF_BASE_ADDRESS
+    isr_handler(_PORTF_BASE_ADDRESS);
+#endif
+#ifdef _PORTG_BASE_ADDRESS
+    isr_handler(_PORTG_BASE_ADDRESS);
+#endif
+
+    mips32r2_isr_end();
+}
+#elif defined(CPU_FAM_PIC32MZ)
+static void isr_handler(uint32_t port_addr)
+{
+    unsigned int pin = 0;
+    for (pin = 0; pin < GPIO_NUMOF; ++pin) {
```

Comment:
Does the PIC32 implement `clz`?
You could save yourself a couple of iterations of that loop if only few bits are set - https://github.com/RIOT-OS/RIOT/blob/master/cpu/cc2538/periph/gpio.c#L216

Reply:
> Does the PIC32 implement clz?

Indeed ! That is a nice trick. I changed the implementation.